### PR TITLE
Refactor of lighting interfaces

### PR DESCRIPTION
**Pull request by shadielhajj**
Implementation of #211.

I think the lighting pipeline looks much cleaner and simpler now.

Let me know if it all looks good to you @patriciogonzalezvivo  and I'll port it to HLSL.

### CODE REVIEW

**Comment by patriciogonzalezvivo** [comment]
oh Nice!! I will check it out tomorrow morning.

Thank you again!

**Comment by patriciogonzalezvivo** [comment]
Super nice cleaning of all the redundant functions + new datatypes and helper functions !! <3

File: lighting/pbrLittle.glsl [hunk 1]
```diff
@@ -46,76 +46,60 @@ license:
...
-    float smoothness = 0.95 - saturate(_roughness);
+    float notMetal = 1.0 - mat.metallic;
+    float smoothness = 0.95 - saturate(mat.roughness);
 
     #if defined(LIGHT_SHADOWMAP) && defined(LIGHT_SHADOWMAP_SIZE) && defined(LIGHT_COORD)
-    _shadow *= shadow(LIGHT_SHADOWMAP, vec2(LIGHT_SHADOWMAP_SIZE), (LIGHT_COORD).xy, (LIGHT_COORD).z);
+    float shadow = shadow(LIGHT_SHADOWMAP, vec2(LIGHT_SHADOWMAP_SIZE), (LIGHT_COORD).xy, (LIGHT_COORD).z);
     #elif defined(FNC_RAYMARCH_SOFTSHADOW)
-    _shadow *= raymarchSoftShadow(_position, L);
+    float shadow = raymarchSoftShadow(mat.position, shadingData.L);
```
Comment (patriciogonzalezvivo) [inline lighting/pbrLittle.glsl:71]: A default needs to be added. Ex:

```glsl
#else
float shadow = 1.0;
#endif
```

File: lighting/gooch.glsl [hunk 2]
```diff
@@ -1,12 +1,11 @@
...
 #include "material/normal.glsl"
 #include "material/albedo.glsl"
 #include "material.glsl"
 #include "light/new.glsl"
-
-// #include "diffuse.glsl"
-// #include "specular.glsl"
-// #include "../sample/shadowPCF.glsl"
+#include "specular.glsl"
+#include "diffuse.glsl"
```
Comment (patriciogonzalezvivo) [inline lighting/gooch.glsl:8]: missing dependency
```glsl
#include "reflection.glsl"
```

File: lighting/pbr.glsl [hunk 3]
```diff
@@ -45,65 +46,54 @@ license:
...
+
     {
         #if defined(LIGHT_DIRECTION)
         LightDirectional L = LightDirectionalNew();
-        lightResolve(diffuseColor, specularColor, M, L, lightDiffuse, lightSpecular);
         #elif defined(LIGHT_POSITION)
         LightPoint L = LightPointNew();
-        lightResolve(diffuseColor, specularColor, M, L, lightDiffuse, lightSpecular);
         #endif
+        lightResolve(L, mat, shadingData);
```
Comment (patriciogonzalezvivo) [inline lighting/pbr.glsl:91]: This fails if `LIGHT_DIRECTION` and  `LIGHT_POSITION` are not defined

suggestion
```glsl
        #if defined(LIGHT_DIRECTION)
        LightDirectional L = LightDirectionalNew();
        lightResolve(L, mat, shadingData);

        #elif defined(LIGHT_POSITION)
        LightPoint L = LightPointNew();
        lightResolve(L, mat, shadingData);
        #endif
```

File: lighting/light/pointEvaluate.glsl [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+*/
+
+#include "../specular.glsl"
+#include "../diffuse.glsl"
+#include "falloff.glsl"
+
+#ifndef FNC_LIGHT_POINT
+#define FNC_LIGHT_POINT
+
+void lightPointEvaluate(LightPoint L, Material mat, ShadingData shadingData) {
```
Comment (patriciogonzalezvivo) [inline lighting/light/pointEvaluate.glsl:24]: `shadingData` doesn't return. Suggestion

```glsl

void lightPointEvaluate(LightPoint L, Material mat, inout ShadingData shadingData)
```

**Comment by patriciogonzalezvivo** [comment]
Getting the following visual regressions

image

image

image

image

image

image

image

image

**Comment by shadielhajj** [comment]
> Getting the following visual regressions

On it. @patriciogonzalezvivo is there a script to get these nice side-by-side tests?

File: lighting/gooch.glsl (hunk 2 above)
Comment (shadielhajj) [reply lighting/gooch.glsl:8]: Thanks for the catch. Fixed.

File: lighting/light/pointEvaluate.glsl (hunk 4 above)
Comment (shadielhajj) [reply lighting/light/pointEvaluate.glsl:24]: Good catch. Fixed.

File: lighting/pbr.glsl (hunk 3 above)
Comment (shadielhajj) [reply lighting/pbr.glsl:91]: Absolutely. Thanks.

**Comment by patriciogonzalezvivo** [comment]
I generate new screenshots locally and use vs code git tab to check on image that had change. I'm improving the pipeline of lygia_example to generate the makefile, readme and benchmark tests.

**Comment by shadielhajj** [comment]
@patriciogonzalezvivo Should be all working now:

* `pbrLittle` examples: FIXED
* `pbrClearCoat`examples: FIXED
* AOV example. `lighting_raymarching_return.frag` needs to be updated to use the `RAYMARCH_AOV` define instead of the older `RAYMARCH_RETURN` see #206 
* `ssr` example this also needs modification on the client side to account for the new interface. In `lighting_ssr.frag` could to the following:`
```glsl
 #if defined(SCENE_BUFFER_0)
    // SPECULAR BUFFER
    ShadingData shadingData = shadingDataNew();
    shadingData.L = normalize(LIGHT_POSITION - (SURFACE_POSITION).xyz);
    shadingData.N = normalize(material.normal);
    shadingData.V = normalize(CAMERA_POSITION - (SURFACE_POSITION).xyz);
    shadingData.roughness =  material.roughness;
    shadingData.linearRoughness =  material.roughness;
    color += saturate(specular(shadingData));
    float spec =  saturate(.95 - material.roughness * 0.5);
    color.rgb *= 1.0-saturate( (powFast(spec, 4.) * 0.8 + 1.6 * (1.0-material.metallic)) );
    #else
``` 

Let me know if there's any visual regressions left, but I think we should be good now.

**Comment by patriciogonzalezvivo** [comment]
There still some weird specular/reflections

image

image

image

and lack of depth information on raymarching

image

**Comment by patriciogonzalezvivo** [comment]
Oh, haven't notice that change. What `RAYMARCH_AOV` stands for? Could we leave `RAYMARCH_RETURN` instead?

**Comment by shadielhajj** [comment]
@patriciogonzalezvivo AOV (Arbitrary Output Variables) is the name used in pro rendering packages (Unity, Redshift, Octane, ...), and is quite standard in the industry. But I can leave it as `RAYMARCH_RETURN ` if you prefer. Btw this change was done and documented in #206.

**Comment by shadielhajj** [comment]
@patriciogonzalezvivo I think it's all fixed now! 😊
HLSL port should also be ready.
Take a look when you have a sec and please merge it if looks all good.

**Comment by patriciogonzalezvivo** [comment]
image

image

I still see some differences here. The spheres on the on the back are a bit dull

Also:

image

**Comment by shadielhajj** [comment]
@patriciogonzalezvivo Ah yes, yell spotted for the back sphere, I'll investigate.

Regarding the SSR example, as I mentioned in a previous comment, this one requires a client-side change (ie. a Patricio amend), since the interface of the `specular` function has changed. In `lighting_ssr.frag` you could do the following:
```glsl
 #if defined(SCENE_BUFFER_0)
    // SPECULAR BUFFER
    ShadingData shadingData = shadingDataNew();
    shadingData.L = normalize(LIGHT_POSITION - (SURFACE_POSITION).xyz);
    shadingData.N = normalize(material.normal);
    shadingData.V = normalize(CAMERA_POSITION - (SURFACE_POSITION).xyz);
    shadingData.roughness =  material.roughness;
    shadingData.linearRoughness =  material.roughness;
    color += saturate(specular(shadingData));
    float spec =  saturate(.95 - material.roughness * 0.5);
    color.rgb *= 1.0-saturate( (powFast(spec, 4.) * 0.8 + 1.6 * (1.0-material.metallic)) );
    #else
```

**Comment by shadielhajj** [comment]
@patriciogonzalezvivo All the spheres look identical now in the clear coat examples. Hopefully, I didn't forget anything this time.
Thanks for your help testing the refactor, btw. It's hard to not miss anything on this scale.

**Comment by patriciogonzalezvivo** [comment]
great work! thank you!

**Comment by patriciogonzalezvivo** [comment]
We hit an important regression that manifest only on PLATFORM_WEBGL after this merge. 

```
';' : syntax error
```

image

Do you mind helping me to find it?

**Comment by shadielhajj** [comment]
@patriciogonzalezvivo sure! How can I test this? I can't seem to be able to reproduce it with my toolset. And the error message is a bit vague.

**Comment by shadielhajj** [comment]
@patriciogonzalezvivo One thing that comes to mind without being able to reproduce the error, are the defines in `specular`and `diffuse`. Might be worth trying different functions.

```glsl
#ifndef DIFFUSE_FNC 
#if defined(TARGET_MOBILE) || defined(PLATFORM_RPI) || defined(PLATFORM_WEBGL)
#define DIFFUSE_FNC diffuseLambert
#else
#define DIFFUSE_FNC diffuseOrenNayar
#endif  
#endif
``` 

```glsl
#ifndef SPECULAR_FNC
#if defined(TARGET_MOBILE) || defined(PLATFORM_RPI) || defined(PLATFORM_WEBGL)
#define SPECULAR_FNC specularBlinnPhongRoughness
#else
#define SPECULAR_FNC specularCookTorrance
#endif  
#endif
```

**Comment by patriciogonzalezvivo** [comment]
Found it! Was pretty hard to spot and only manifesting on WebGL 1.0 : )

https://github.com/patriciogonzalezvivo/lygia/commit/556a1e31fd80f21406f370e4eaf7ca7a99c20362

Man... WebGL 1.0 is such a pain in the but.

Great work! That was a massive rearch!

**Comment by patriciogonzalezvivo** [comment]
Question, finding the bug I notice the define in `raymarch.?lsl` is `#define ENG_RAYMARCH` it's ok if we revert this one to `#define FNC_RAYMARCH` for consistency?

**Comment by shadielhajj** [comment]
@patriciogonzalezvivo Wow! Well spotted 😅😅😅 and sorry about that!
`FNC_RAYMARCH` is totally fine! `Not sure where this `ENG_` prefix comes from.

**Pull request merged**